src/eytzinger_benchmark.zig

```diff
@@ -0,0 +1,319 @@
+
+    inline for (kv_types) |kv| {
+        inline for (values_per_page) |values_count, v| {
+            inline for (summary_sizes[v]) |keys_count| {
+                try run_benchmark(.{
+                    .blob_size = blob_size,
+                    .key_size = kv.key_size,
+                    .value_size = kv.value_size,
+                    .keys_count = keys_count,
+                    .values_count = values_count,
+                    .searches = searches,
+                }, blob, &prng.random);
+            }
+        }
+    }
+}
+
+fn run_benchmark(comptime layout: Layout, blob: []u8, random: *std.rand.Random) !void {
+    assert(blob.len == layout.blob_size);
+    const Eytzinger = eytzinger(layout.keys_count - 1, layout.values_count);
+    const Val = Value(layout);
+    const Key = Val.Key;
+    const Page = struct {
+        summary: [layout.keys_count]Key,
+        values: [layout.values_count]Val,
+    };
+    const page_count = layout.blob_size / @sizeOf(Page);
+
+    // Search pages and keys in random order.
+    var page_picker = shuffled_index(page_count, random);
+    var value_picker = shuffled_index(layout.values_count, random);
+
+    // Generate 1GiB worth of 24KiB pages.
+    var blob_alloc = std.heap.FixedBufferAllocator.init(blob);
+    var pages = try blob_alloc.allocator.alloc(Page, page_count);
+    random.bytes(std.mem.sliceAsBytes(pages));
+    for (pages) |*page| {
+        for (page.values) |*value, i| value.key = i;
+        Eytzinger.layout_from_keys_or_values(
+            Key, Val, Val.key_from_value, Val.max_key,
+            &page.values, &page.summary);
+    }
+
+    const stdout = std.io.getStdOut().writer();
+    {
+        var benchmark = try Benchmark.begin();
+        var i: usize = 0;
+        var v: usize = 0;
+        while (i < layout.searches) : (i += 1) {
+            const page_index = page_picker[i % page_picker.len];
+            const target = value_picker[v % value_picker.len];
+            const page = pages[page_index];
+            const bounds = Eytzinger.search_values(Key, Val, Val.key_compare, &page.summary, &page.values, target);
+            assert(bounds.len != 0);
+            const hit = if (bounds.len == 1) bounds[0]
+                else bounds[binary_search(Key, Val, Val.key_from_value, Val.key_compare, bounds, target)];
+
+            assert(hit.key == target);
+            if (i % pages.len == 0) v += 1;
+        }
+
+        const result = try benchmark.end(layout.searches);
+        try stdout.print(body_fmt, .{
+            layout.key_size,
+            layout.value_size,
+            layout.keys_count,
+            layout.values_count,
+            "E",
+            "B",
+            result.wall_time,
+            result.utime,
+            result.cpu_cycles,
+            result.instructions,
+            result.cache_references,
+            result.cache_misses,
+            result.branch_misses,
+        });
+    }
+
+    {
```

Comment:
It might be interesting to experiment with sometimes running the Binary Search benchmark first, other times the Eytzinger benchmark (alternating their respective order), just to see what happens.

Reply:
No impact that I can see, but the results are posted to the google sheet.